File: sdk/android/src/main/kotlin/servify/sdk/android/core/HandoffStateMachine.kt
package servify.sdk.android.core

/**
 * 转人工状态机（转移表与 PROTOCOL.md §4.2 逐条对应，事件全部为真实帧）：
 *
 * ```
 * ai_answering  ──(transfer_notification)──> agent_chatting
 * ai_answering  ──(waiting_notification)───> waiting_human
 * waiting_human ──(transfer_notification)──> agent_chatting
 * ```
 *
 * 非法转移被拒绝并返回 false——状态机转移表必须穷举（策划文档 M1 验收⑤），
 * 收到语义上不可能的帧序列时显式暴露而不是静默吞掉。
 */
class HandoffStateMachine(initial: HandoffState = HandoffState.AiAnswering) {

    var state: HandoffState = initial
        private set

    /** 已发生的合法转移记录（from → to），fixtures 回放断言用。 */
    val transitions = mutableListOf<Pair<HandoffState, HandoffState>>()

    fun onTransferNotification(): Boolean {
        val legal = state == HandoffState.AiAnswering || state == HandoffState.WaitingHuman
        if (legal) transitionTo(HandoffState.AgentChatting)
        return legal
    }

    fun onWaitingNotification(): Boolean {
        val legal = state == HandoffState.AiAnswering
        if (legal) transitionTo(HandoffState.WaitingHuman)
        return legal
    }

    private fun transitionTo(next: HandoffState) {
        transitions += state to next
        state = next
    }
}

sealed interface HandoffState {
    data object AiAnswering : HandoffState
    data object WaitingHuman : HandoffState
    data object AgentChatting : HandoffState
    data object Closed : HandoffState
}

File: sdk/ios/Sources/ServifyKit/Core/HandoffStateMachine.swift
import Foundation

/// 转人工状态机（转移表与 PROTOCOL.md §4.2 逐条对应，事件全部为真实帧）：
///
/// ```
/// ai_answering  ──(transfer_notification)──> agent_chatting
/// ai_answering  ──(waiting_notification)───> waiting_human
/// waiting_human ──(transfer_notification)──> agent_chatting
/// ```
///
/// 非法转移被拒绝并返回 false——状态机转移表必须穷举（策划文档验收⑤），收到语义上
/// 不可能的帧序列时显式暴露而不是静默吞掉。Kotlin 镜像：shared/core HandoffStateMachine.kt。
public final class HandoffStateMachine {

    public private(set) var state: HandoffState

    /// 已发生的合法转移记录（from → to），fixtures 回放断言用。
    public private(set) var transitions: [HandoffTransition] = []

    public init(initial: HandoffState = .aiAnswering) {
        state = initial
    }

    @discardableResult
    public func onTransferNotification() -> Bool {
        let legal = state == .aiAnswering || state == .waitingHuman
        if legal { transitionTo(.agentChatting) }
        return legal
    }

    @discardableResult
    public func onWaitingNotification() -> Bool {
        let legal = state == .aiAnswering
        if legal { transitionTo(.waitingHuman) }
        return legal
    }

    private func transitionTo(_ next: HandoffState) {
        transitions.append(HandoffTransition(from: state, to: next))
        state = next
    }
}

public enum HandoffState: Equatable, Hashable {
    case aiAnswering
    case waitingHuman
    case agentChatting
    /// 设计在案、V1 无入边：进入依赖补拉语义（服务端配套项落地前不接线，防死代码）。
    case closed
}

public struct HandoffTransition: Equatable {
    public let from: HandoffState
    public let to: HandoffState

    public init(from: HandoffState, to: HandoffState) {
        self.from = from
        self.to = to
    }
}
